messaging-spec.md

```diff
@@ -0,0 +1,351 @@
+
+Conclusion: Since blockspace is consumed both when the dispatchable is initially called and when the callback executes, we effectively pay for the callback's weight twice, once in Block 1 and again in Block 2.
+
+
+A solution to this is deducting the weight of the callback + callback execution in block 1.
+Since the weight is not actually being used (since we are precharging), this can be considered safe (but a little hacky).
+We can acheive this using `TransactionExtension` specifically the `StorageWeightReclaim` type.
+
+#### On XCMs handling of weight for responses.
+
+TODO!("Does XCM charge the weight of the runtime call when it handles a response?")
+
+
+### Response Hooks
+For ismp we have the `IsmpModule` trait which is called when ismp-pallet receives a response, timeout or accepts a request. 
+We also have the corrosponding `IsmpModuleWeight` which we can use for the pre execution charge.
+
+For xcm its a little different, we must define an extrinsic that xcm-pallet calls as its notify parameter. 
+Then we benchmark the response extrinsic like normal.
+
+Both of these are dependant on the CallbackExecutor.
+
+### XCM and Queries
+We need to tell pallet-xcm to expect a new query regardless of which method we use to generate message ids. 
+For this we can use: 
+```rust
+	/// Attempt to create a new query ID and register it as a query that is yet to respond, and
+	/// which will call a dispatchable when a response happens.
+	pub fn new_notify_query(
+		responder: impl Into<Location>,
+		notify: impl Into<<T as Config>::RuntimeCall>,
+		timeout: BlockNumberFor<T>,
+		match_querier: impl Into<Location>,
+	) -> u64 {
+        // snip 
+    }
+```
+Unfortunately the QueryHandler trait doesnt contain the notify parameter hence we have to use a wrapper trait: 
+```rust
+pub trait NotifyQueryHandler<T: Config> {
+	/// Attempt to create a new query ID and register it as a query that is yet to respond, and
+	/// which will call a dispatchable when a response happens.
+	fn new_notify_query(
+		responder: impl Into<Location>,
+		notify: Call<T>,
+		timeout: BlockNumberOf<T>,
+		match_querier: impl Into<Location>,
+	) -> u64;
+}
+```
+
+This returns us a XCM query_id, not to be confused with our message_id, which we can use to match to a request and therefore the registered call back.
+When pallet-xcm recieves the response it will decode the runtime call and call our extrinsic for response handling.
+
+### ISMP and Queries
+We can use the `IsmpDispatcher` to dispatch all outgoing calls.
+
+### Message IDs
+Message ids are used as a key for outgoing requests, they can be queried to find the status of a message.
+A message id can either be specified by the runtime or the contract.
+
+If the message id is specified by the runtime, the runtime will have to return that data to the contract to emit or emit an event itself. The event emitted must contain data that can relate the request to a response.
```

Comment:
I think we should choose one or the other (contract manages its own request IDs, or runtime manages request IDs). But not both as that adds complexity, collision checking, etc. 

- Contract passes request id seems to be simplest route for our pallet. 
- Runtime returns request id
    - Contract can still create their own system to map their requests to the pallets request ID
    - We would have to modify the API to support executing a dispatchable, and then returning a request ID. This may become too complicated

Reply:
collision checking will happen by default (through seeing if the message exists before creating one).
Prevention of collision perhaps is unnecessary.

I will change to an active voice.